srml/contract/src/gas.rs

```diff
@@ -48,17 +90,42 @@ impl<T: Trait> GasMeter<T> {
 			limit: gas_limit,
 			gas_left: gas_limit,
 			gas_price,
+			#[cfg(test)]
+			tokens: Vec::new(),
 		}
 	}
 
 	/// Account for used gas.
 	///
+	/// Amount is calculated by the given `token`. If `token::calculate_amount` returns
+	/// `None` then all available gas is consumed and `OutOfGas` is returned.
+	///
 	/// Returns `OutOfGas` if there is not enough gas or addition of the specified
 	/// amount of gas has lead to overflow. On success returns `Proceed`.
 	///
-	/// NOTE that `amount` is always consumed, i.e. if there is not enough gas
+	/// NOTE that amount is always consumed, i.e. if there is not enough gas
 	/// then the counter will be set to zero.
-	pub fn charge(&mut self, amount: T::Gas) -> GasMeterResult {
+	#[inline]
+	pub fn charge<Tok: Token<T>>(
+		&mut self,
+		metadata: &Tok::Metadata,
+		token: Tok,
+	) -> GasMeterResult {
+		#[cfg(test)]
+		{
+			// Unconditionally add the token to the storage.
+			let erased_tok = ErasedToken {
```

Comment:
Do we really want to add test code into production code? Can we not mock `GasMeter` in some way?

Reply:
I think we would not need to go the full generic way. We could try the following:
```
struct GasMeterImpl {}

#[cfg(test)]
struct GasMeterMock {}

#[cfg(test)]
pub type GasMeter = GasMeterMock {};

#[cfg(not(test))]
pub type GasMeter = GasMeterImpl;
```

By doing it this way, we would not require any trait. We would just require to replicate function signatures. And also only the ones we require in tests.

Reply:
Yeah that should work! But tbh I still feel that it doesn't outweights the simple `#[cfg(test)]` solution. In tests we still require the logic of the production version of `GasMeter`. Implemented naively this will end up with duplication and duplication deminishes the value of tests since tests now testing not the real implementation but the copy of it... 

I think we can try to wrap `GasMeterImpl ` with `GasMeterMock`, but it will still involve some amount of duplication.

Maybe we should make a step back and mb find an another way? Why exactly do you think the current solution is bad? What problems could arise with the current approach?

Reply:
I would just have used `GasMeterImpl` in `GasMeterMock` and forwared all calls.

In my opinion testing should be done in such a way, that you take your struct as blackbox and just test that when you give `x` in you get `y` and not `z`. The current test code is also not that much, but it could start to make reading the real implementation much more complex.

Reply:
I see what you mean, yes! I can imagine where this can end up, but for now it is really a small part of the code.

How about this compromise: we just leave a comment right there, which says that if one try to pile some other debugging code then he/she/they should consider type-alias approach. How does this sound for you?